### DISCUSSION TITLE
GitHub Secret/ENV var population in NextJS server actions WITH useActionState hook

### DISCUSSION THREAD

**Discussion by bakirgdev** [vercel/next.js#74446] [Help] [answered]
### PROBLEM
I came to a conclusion after many days of frustration and testing that imported ENV variables (from github secrets) that are used in server action via useActionState hook seemingly MUST be public variables, i.e. with `NEXT_PUBLIC_`

### BEHAVIOUR
BEFORE there was no `NEXT_PUBLIC_` prefix to my env variables in code (see files below). but after some research i added the `NEXT_PUBLIC_` prefix, and i had no population errors. in the old case i got 404 on Telegram api call hinting that the bot token and chatid were not found since they were not populated or accessible. i mistook the error was on github side and my repo not feeding the ENV vars somehow.

during local development i used .env.local and inside the necessary variables WITHOUT `NEXT_PUBLIC_` prefix -- and it worked suprisingly!

### NEED EXPLANATION
in summary, i need you to view the files i provided and give me a brief explanation: is this the expected behaviour? since the main point of server actions is that they are ran on the server side thus not needing `NEXT_PUBLIC_` prefix in ENV vars. but i get to understand that by using the useActionState hook somehow env vars need to be public in server actions?! simply makes no sense in my head

i reviewed the code on the page (https://bakir.dev/links) where the form is, and have not found via basic Ctrl+F search any instances of "NEXT_PUBLIC"

### NOTE
in advance i am sorry for not reading any related docs, issues or similar that are related to this, i just want to engage more with this awesome community! 

### RELATED FILES !
please review these files
- FORM COMPONENT WITH useActionState HOOK: https://github.com/BakirGracic/bakir-dev/blob/main/components/ContactForm.tsx
- SERVER ACTION FILE: https://github.com/BakirGracic/bakir-dev/blob/main/actions/contact-form.ts
- WORKFLOW FILE WHERE VARS ARE POPULATED: https://github.com/BakirGracic/bakir-dev/blob/main/.github/workflows/nextjs-cd.yml

### Additional information

```js
https://github.com/BakirGracic/bakir-dev
```

### Example

https://github.com/BakirGracic/bakir-dev

**Comment by icyJoseph** [comment, marked as answer]
Do you mean that, your env vars, were available during GitHub Actions build time, but not on the runtime (aka not GitHub actions anymore)? 

That makes sense... NEXT_PUBLIC vars are inlined during build time, and the others (without NEXT_PUBLIC), UNLESS the route is pre-rendered, are left as "open", so that you can replace them with whatever your system has at start time. Again, unless the route is pre-rendered, on which case they pick up the value from the build time.

Test it out here, https://stackblitz.com/edit/nextjs-racxberv?file=app%2Fpage.tsx,app%2Flayout.tsx,package.json - just `npm install` and then do:

```console
npm run build && npm run start
```

Comment out `export const dynamic = 'force-dynamic';`, to make it pre-rendered, and see the results of, `npm run build && npm run start`, once again.

**Reply by bakirgdev** [reply]
I really complicated my question :/

My goal was to conceal 2 tokens in GitHub Secrets, and import them in my app securely in Actions workflow where I build and start my app (CD).

So far I 'inject' ENV variables at BUILD TIME with `env:` YAML syntax then run the build command.

The problem is that `TELEGRAM_BOT_TOKEN` and `TELEGRAM_CHAT_ID` ENV variables in production/live-site were not retrievable/accessible in code, using the method above.

I implement those 2 variables in Server Actions (`'use server'`) to manage a form.

I thought my GitHub repo was a problem, OFC no.
Then maybe `useActionState` converts the Server Action to be somehow executed on the client, also didn't make sense.
Then I was confused when I converted my variables to client-side (using `NEXT_PUBLIC_` prefix) and that worked, but knowing that Server Actions are exec on server and I could not find my variable in client side code

Your example is very simple and neat. Now I know there are ways to inject vars at build time and on start.

I implemented your suggestion (to inject ENV vars as default server-side ON RUN TIME) - and it worked...

Seems like I don't know enough regarding Node ENV variables, rookie mistake, will improve!

Thanks @icyJoseph
